[ERROR] agent_fox.engine.session_lifecycle: Harvest failed

/Users/candlekeep/devel/workspace/fox-benchmark
✔ 01_project_scaffold:2 done (2m 54s)
⠙ [01_project_scaffold:3] thinking...[ERROR] agent_fox.engine.session_lifecycle: Harvest failed for 01_project_scaffold:3 after successful session: Merge of 'feature/01_project_scaffold/3' failed: Auto-merging .specs/01_project_scaffold/tasks.md
Auto-merging Makefile
CONFLICT (add/add): Merge conflict in Makefile
Auto-merging build_test.go
CONFLICT (add/add): Merge conflict in build_test.go
Auto-merging cmd/fox-paste-server/main.go
CONFLICT (add/add): Merge conflict in cmd/fox-paste-server/main.go
Auto-merging go.mod
CONFLICT (add/add): Merge conflict in go.mod
Auto-merging internal/config/config.go
CONFLICT (add/add): Merge conflict in internal/config/config.go
Auto-merging internal/config/config_test.go
CONFLICT (add/add): Merge conflict in internal/config/config_test.go
Auto-merging internal/database/database.go
CONFLICT (add/add): Merge conflict in internal/database/database.go
Auto-merging internal/database/database_test.go
CONFLICT (add/add): Merge conflict in internal/database/database_test.go
Auto-merging internal/logging/logging.go
CONFLICT (add/add): Merge conflict in internal/logging/logging.go
Auto-merging internal/logging/logging_test.go
CONFLICT (add/add): Merge conflict in internal/logging/logging_test.go
Auto-merging internal/migrate/migrate.go
CONFLICT (add/add): Merge conflict in internal/migrate/migrate.go
Auto-merging internal/migrate/migrate_test.go
CONFLICT (add/add): Merge conflict in internal/migrate/migrate_test.go
Auto-merging internal/server/lifecycle_test.go
CONFLICT (add/add): Merge conflict in internal/server/lifecycle_test.go
Auto-merging internal/server/server.go
CONFLICT (add/add): Merge conflict in internal/server/server.go
Auto-merging internal/server/server_test.go
CONFLICT (add/add): Merge conflict in internal/server/server_test.go
Automatic merge failed; fix conflicts and then commit the result.

--- a/agent_fox/workspace/git.py
+++ b/agent_fox/workspace/git.py
@@ -162,17 +162,29 @@ async def merge_fast_forward(
 async def merge_commit(
     repo_path: Path,
     branch: str,
+    *,
+    strategy_option: str | None = None,
 ) -> None:
     """Merge branch into HEAD with a merge commit.
 
     Falls back to a regular (non-fast-forward) merge when a
     fast-forward is not possible.
 
+    Args:
+        strategy_option: If set, passed as ``-X {value}`` to git merge
+            (e.g. ``"theirs"`` to auto-resolve conflicts by preferring
+            the incoming branch).
+
     Raises:
         IntegrationError: If the merge fails (conflicts).
     """
+    cmd = ["merge", "--no-edit"]
+    if strategy_option:
+        cmd.extend(["-X", strategy_option])
+    cmd.append(branch)
+
     returncode, stdout, stderr = await run_git(
-        ["merge", "--no-edit", branch],
+        cmd,
         cwd=repo_path,
         check=False,
     )

--- a/agent_fox/workspace/harvester.py
+++ b/agent_fox/workspace/harvester.py
@@ -86,14 +86,24 @@ async def harvest(
     except IntegrationError:
         # Rebase failed (conflicts) — abort and fall back to merge commit
         logger.info(
-            "Rebase of '%s' onto '%s' had conflicts, "
-            "falling back to merge commit",
+            "Rebase of '%s' onto '%s' had conflicts, falling back to merge commit",
             workspace.branch,
             dev_branch,
         )
         await abort_rebase(workspace.path)
         # Step 5: Fall back to regular merge (03-REQ-7.E1)
-        await merge_commit(repo_root, workspace.branch)
+        try:
+            await merge_commit(repo_root, workspace.branch)
+        except IntegrationError:
+            # Step 6: Auto-resolve conflicts preferring the feature branch.
+            # This handles add/add conflicts from parallel sessions where
+            # multiple task groups independently create the same files.
+            logger.info(
+                "Merge commit of '%s' failed, retrying with "
+                "-X theirs to auto-resolve conflicts",
+                workspace.branch,
+            )
+            await merge_commit(repo_root, workspace.branch, strategy_option="theirs")
         changed_files = await get_changed_files(
             repo_root,
             workspace.branch,

--- a/tests/unit/workspace/test_harvester.py
+++ b/tests/unit/workspace/test_harvester.py
@@ -12,7 +12,6 @@ from pathlib import Path
 
 import pytest
 
-from agent_fox.core.errors import IntegrationError
 from agent_fox.workspace.harvester import harvest
 from agent_fox.workspace.worktree import create_worktree
 
@@ -157,25 +156,26 @@ class TestHarvesterNoCommits:
         assert develop_tip_after == develop_tip_before
 
 
-class TestHarvesterUnresolvableConflict:
-    """TS-03-E6: Harvester raises IntegrationError on unresolvable conflict."""
+class TestHarvesterConflictAutoResolve:
+    """Harvester auto-resolves add/add conflicts preferring feature branch."""
 
     @pytest.mark.asyncio
-    async def test_conflict_raises_integration_error(
+    async def test_add_add_conflict_resolved_with_theirs(
         self,
         tmp_worktree_repo: Path,
     ) -> None:
-        """An unresolvable merge conflict raises IntegrationError."""
+        """When both branches add the same file with different content,
+        the harvester auto-resolves by keeping the feature branch version."""
         ws = await create_worktree(tmp_worktree_repo, "test_spec", 1)
 
-        # Modify the same file on the feature branch
+        # Add a file on the feature branch
         add_commit_to_branch(
             ws.path,
             "shared.py",
             "feature content\n",
         )
 
-        # Modify the same file on develop (creates conflict)
+        # Add the SAME file with DIFFERENT content on develop
         subprocess.run(
             ["git", "checkout", "develop"],
             cwd=tmp_worktree_repo,
@@ -188,69 +188,89 @@ class TestHarvesterUnresolvableConflict:
             "develop content\n",
         )
 
-        with pytest.raises(IntegrationError):
-            await harvest(tmp_worktree_repo, ws)
+        # Harvest should succeed (no IntegrationError)
+        files = await harvest(tmp_worktree_repo, ws)
+        assert isinstance(files, list)
 
-    @pytest.mark.asyncio
-    async def test_conflict_error_includes_details(
-        self,
-        tmp_worktree_repo: Path,
-    ) -> None:
-        """IntegrationError message includes conflict file details."""
-        ws = await create_worktree(tmp_worktree_repo, "test_spec", 1)
-
-        add_commit_to_branch(
-            ws.path,
-            "shared.py",
-            "feature content\n",
-        )
-
+        # The feature branch content should win on develop
         subprocess.run(
             ["git", "checkout", "develop"],
             cwd=tmp_worktree_repo,
             check=True,
             capture_output=True,
         )
-        add_commit_to_branch(
-            tmp_worktree_repo,
-            "shared.py",
-            "develop content\n",
-        )
-
-        with pytest.raises(IntegrationError, match="shared.py"):
-            await harvest(tmp_worktree_repo, ws)
+        shared = (tmp_worktree_repo / "shared.py").read_text()
+        assert shared == "feature content\n"
 
     @pytest.mark.asyncio
-    async def test_conflict_leaves_develop_unchanged(
+    async def test_parallel_add_add_multiple_files(
         self,
         tmp_worktree_repo: Path,
     ) -> None:
-        """After a conflict, develop remains at its original tip."""
+        """Simulates parallel sessions creating overlapping files —
+        the exact scenario from issue #84."""
         ws = await create_worktree(tmp_worktree_repo, "test_spec", 1)
 
-        add_commit_to_branch(
-            ws.path,
-            "shared.py",
-            "feature content\n",
-        )
+        # Feature branch creates several files (simulating a task group)
+        add_commit_to_branch(ws.path, "Makefile", "feature-makefile\n")
+        add_commit_to_branch(ws.path, "go.mod", "feature-gomod\n")
 
+        # Meanwhile, develop got the same files from another session
         subprocess.run(
             ["git", "checkout", "develop"],
             cwd=tmp_worktree_repo,
             check=True,
             capture_output=True,
         )
-        add_commit_to_branch(
-            tmp_worktree_repo,
-            "shared.py",
-            "develop content\n",
+        add_commit_to_branch(tmp_worktree_repo, "Makefile", "develop-makefile\n")
+        add_commit_to_branch(tmp_worktree_repo, "go.mod", "develop-gomod\n")
+
+        # Harvest should succeed (no IntegrationError)
+        files = await harvest(tmp_worktree_repo, ws)
+        assert isinstance(files, list)
+
+        # Feature branch content should win for both files
+        subprocess.run(
+            ["git", "checkout", "develop"],
+            cwd=tmp_worktree_repo,
+            check=True,
+            capture_output=True,
         )
-        # Capture develop tip after the divergent commit
-        develop_tip_after_commit = get_branch_tip(tmp_worktree_repo, "develop")
+        assert (tmp_worktree_repo / "Makefile").read_text() == "feature-makefile\n"
+        assert (tmp_worktree_repo / "go.mod").read_text() == "feature-gomod\n"
 
-        with pytest.raises(IntegrationError):
-            await harvest(tmp_worktree_repo, ws)
+    @pytest.mark.asyncio
+    async def test_auto_resolve_preserves_non_conflicting_develop_changes(
+        self,
+        tmp_worktree_repo: Path,
+    ) -> None:
+        """Non-conflicting changes from develop are preserved."""
+        ws = await create_worktree(tmp_worktree_repo, "test_spec", 1)
 
-        # Develop should be unchanged from the point after the divergent commit
-        develop_tip_after_harvest = get_branch_tip(tmp_worktree_repo, "develop")
-        assert develop_tip_after_harvest == develop_tip_after_commit
+        # Feature branch creates one file
+        add_commit_to_branch(ws.path, "shared.py", "feature content\n")
+
+        # Develop creates the same file AND a different file
+        subprocess.run(
+            ["git", "checkout", "develop"],
+            cwd=tmp_worktree_repo,
+            check=True,
+            capture_output=True,
+        )
+        add_commit_to_branch(tmp_worktree_repo, "shared.py", "develop content\n")
+        add_commit_to_branch(tmp_worktree_repo, "other.py", "other content\n")
+
+        files = await harvest(tmp_worktree_repo, ws)
+        assert isinstance(files, list)
+
+        # Checkout develop to verify merged content
+        subprocess.run(
+            ["git", "checkout", "develop"],
+            cwd=tmp_worktree_repo,
+            check=True,
+            capture_output=True,
+        )
+        # Feature wins for the conflict
+        assert (tmp_worktree_repo / "shared.py").read_text() == "feature content\n"
+        # Develop's non-conflicting file is preserved
+        assert (tmp_worktree_repo / "other.py").read_text() == "other content\n"

